skyportal/handlers/api/gcn.py

```diff
@@ -0,0 +1,218 @@
+# Inspired by https://github.com/growth-astro/growth-too-marshal/blob/main/growth/too/gcn.py
+
+import datetime
+import os
+import gcn
+import lxml
+import xmlschema
+from urllib.parse import urlparse
+
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.orm import joinedload
+
+from baselayer.app.access import auth_or_token
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    GcnEvent,
+    GcnNotice,
+    GcnTag,
+    Localization,
+)
+from ...utils.gcn import get_dateobs, get_tags, get_skymap, get_contour
+
+
+class GcnHandler(BaseHandler):
+    @auth_or_token
+    def put(self):
+        """
+        ---
+        description: Ingest GCN xml file
+        tags:
+          - gcnevents
+          - gcntags
+          - gcnnotices
+          - localizations
+        requestBody:
+          content:
+            application/json:
+              schema: GcnHandlerPut
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+        data = self.get_json()
+        payload = data['xml']
+
+        schema = f'{os.path.dirname(__file__)}/../../utils/schema/VOEvent-v2.0.xsd'
+        voevent_schema = xmlschema.XMLSchema(schema)
+        if voevent_schema.is_valid(payload):
+            root = lxml.etree.fromstring(payload.encode('ascii'))
+        else:
+            raise Exception("xml file is not valid VOEvent")
+
+        dateobs = get_dateobs(root)
+
+        try:
+            event = GcnEvent.query.filter_by(dateobs=dateobs).one()
+        except NoResultFound:
+            DBSession().add(
+                GcnEvent(dateobs=dateobs, sent_by_id=self.associated_user_object.id)
+            )
+            self.verify_and_commit()
+            event = GcnEvent.query.filter_by(dateobs=dateobs).one()
```

Comment:
```suggestion
            event = GcnEvent(dateobs=dateobs, sent_by_id=self.associated_user_object.id)
            DBSession().add(event)
```
This is the pattern I used to get rid of the early commit with the localization below, and I think can be applied here so we're not committing mid-handler. Sorry to bring it up again haha, but I would really prefer if we could just commit once at the end if we can- it simplifies rolling back transactions if anything goes wrong a lot :)